include/experimental/__p1673_bits/blas1_matrix_frob_norm.hpp

```diff
@@ -0,0 +1,126 @@
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef LINALG_INCLUDE_EXPERIMENTAL___P1673_BITS_BLAS1_MATRIX_FROB_NORM_HPP_
+#define LINALG_INCLUDE_EXPERIMENTAL___P1673_BITS_BLAS1_MATRIX_FROB_NORM_HPP_
+
+#include <cmath>
+#include <cstdlib>
+
+namespace std {
+namespace experimental {
+inline namespace __p1673_version_0 {
+namespace linalg {
+
+template<class in_matrix_t,
+         class Scalar>
+Scalar matrix_frob_norm(in_matrix_t A,
+                        Scalar init)
+{
+  using std::abs;
+  using std::sqrt;
+
+  // Handle special cases.
+  auto result = init;
+  if (A.extent(0) == 0 || A.extent(1) == 0) {
+    return result;
+  }
+  else if(A.extent(0) == ptrdiff_t(1) && A.extent(1) == ptrdiff_t(1)) {
+    result += abs(A(0, 0));
+    return result;
+  }
+
+  // Rescaling avoids unwarranted overflow or underflow.
+  Scalar scale = 0.0;
+  Scalar ssq = 1.0;
+  for (ptrdiff_t i = 0; i < A.extent(0); ++i) {
+    for (ptrdiff_t j = 0; j < A.extent(1); ++j) {
+      const auto absaij = abs(A(i,j));
+      if (absaij != 0.0) {
+        const auto quotient = scale / absaij;
+        if (scale < absaij) {
+          ssq = Scalar(1.0) + ssq * quotient * quotient;
+          scale = absaij;
+        }
+        else {
+          ssq = ssq + quotient * quotient;
+        }
+      }
+    }
+  }
+  result += scale * sqrt(ssq);
+  return result;
+}
+
+template<class ExecutionPolicy,
+         class in_matrix_t,
+         class Scalar>
+Scalar matrix_frob_norm(ExecutionPolicy&& exec,
+                        in_matrix_t A,
+                        Scalar init)
+{
+  return matrix_frob_norm(A, init);
+}
+
+// TODO: Implement auto functions
```

Comment:
The proposal explains that the return type is `decltype(abs(A(0,0)) * abs(A(0,0)))`.  The resulting function would look like this:
```
template<class in_matrix_t>
auto matrix_frob_norm(in_matrix_t A) -> decltype(abs(A(0,0)) * abs(A(0,0)))
{
  using T = decltype(abs(A(0,0)) * abs(A(0,0)));
  return matrix_frob_norm(A, T{});
}
```